### DISCUSSION TITLE
Generate VEX documents

### DISCUSSION THREAD

**Discussion by dervoeti** [SecObserve/SecObserve#1106] [Ideas]
When exporting vulnerability assessments as VEX, a justification is recommended if the status is set to "Not affected". A list of possible justifications was [published by CISA](https://www.cisa.gov/sites/default/files/publications/VEX_Status_Justification_Jun22.pdf), both [OpenVEX](https://github.com/openvex/spec/blob/main/OPENVEX-SPEC.md#status-justifications) and CSAF adopted this list.

Suggestion: Add a new field "justification" to the `Observation` model and define the list of possible options according to the CISA document. Add a dropdown for this to the "Observation Assessment" dialog and only show it if the status "Not affected" is selected.
I could try to implement this, if you think this makes sense.

**Comment by StefanFl** [comment]
Exporting VEX documents will be a large task, which we have to implement step by step. To start it, I analyzed the OpenVEX and CSAF (VEX profile) specifications against SecObserve's attributes:

* https://drive.google.com/file/d/14waI-qknAAia7p0KJP215NtYhZXCc2Ge/view?usp=sharing is an entity-relationship diagram showing the structures
* https://docs.google.com/spreadsheets/d/18LDZ4kZsUsJBfahTEx12yV3Ce_9Y2Gp4tESTkOdmJhE/edit?usp=sharing lists all the necessary attributes. Green lines are already supported by SecObserve, yellow lines mark where work needs to be done.

Steps could be:

1. Add the justification field to the Observation as described above. I would name it `vex_justification`. @dervoeti, it would be great if you want to do that.
2. Add the fields for `purl` and `cpe23` to the product. @dervoeti Maybe together with the justification field?
3. Add a `VEX` feature flag to SecObserve, so that the new things are only visible for developers. I already have an idea how to do it and will implement it as the next thing.
4. Implement the functionality to generate VEX documents in the backend. This will be several steps and include most likely 1 or 2 new tables. I would add a new Django app `application.vex` for it.
5. Implement the frontend

Steps 4. and 5. will be quite large and be designed later.

**Comment by StefanFl** [comment]
Feature flag is in development: https://github.com/MaibornWolff/SecObserve/pull/1125

**Comment by StefanFl** [comment]
@dervoeti @lfrancke What is a good strategy to generate VEX documents:

1. Maximum 1 document per `Product` / `Product Group` with all vulnerabilites or a selection of vulnerabilities?
2. Multiple documents per `Product` / `Product Group` with a selection of vulnerabilities?
3. Something else?

**Reply by dervoeti** [reply]
I don't really have practical experience with VEX, so I can only answer this from a theoretical perspective, but I would expect us to generate one VEX statement per vulnerability (or group of similar vulnerabilities). For example: “This is our statement regarding Log4Shell vulnerabilities” with information about all affected products and their versions. So we would start by analyzing existing vulnerabilities and publishing VEX statements about them. In the long term, we plan to continuously scan our products and publish VEX statements for newly discovered vulnerabilities (again, a new vulnerability might affect multiple products).

Perhaps it would be possible to solve all strategies by allowing a workflow like this in SecObserve: Select / Filter out various assessments and generate a VEX file for them. Then it would be possible to filter out "all assessments for this product" or "all assessments for these vulnerability IDs" or even "all assessments done within the last 3 days". So, the central logic would be "take N assessments and create a VEX statement for them".

Again, this is only theoretical, I have neither fully thought this through nor experimented with it yet.

**Reply by StefanFl** [reply]
Good point, haven't thought about starting from vulnerabilities yet. Maybe we need both approaches, generating VEX documents for vulnerabilities and products.

Not so sure about starting with vulnerabilities. We might want to report open vulnerabilities without an assessment.

**Reply by lfrancke** [reply]
This will probably evolve heavily over time.
But yes: I think one published VEX per vulnerability makes sense

I can also see that it might be useful for customers later to just say: Give me all VEX statements for this product.
Basically have "query" like functionality. But that's step...3 or so.

**Reply by dervoeti** [reply]
> We might want to report open vulnerabilities without an assessment.

Without an assessment, the question is how to determine the `status` of the VEX. SecObserve's `current_status` "Open" could be mapped to the `status` "under_investigation".

[dep-scan](https://github.com/owasp-dep-scan/dep-scan) does it like that, it creates a VEX statement with "in_triage" (basically the same as "under_investigation", just the CycloneDX name for it) in the CycloneDX document for each vulnerability it found. So that's the default state of the VEX.

Then, the logic would be "take N observations and create a VEX statement for them". An assessment would be optional and just influence the details of the VEX, but would not be necessary to create the VEX.

As a user, an alternative would be to create a rule that sets all new observations to `current_status` = "In review", if I want to report those vulnerabilities. This way, the user could define this mapping explicitly.

**Reply by StefanFl** [reply]
That is something to consider. Current philosophy of SecObserve is: An observation is treated as relevant/open, until someone decides differently. Should be a good default, but maybe not ideal for creating a VEX document. 

I like your suggestion with the rule, that gives a project a way to decide themselves.

**Comment by StefanFl** [comment]
The branch https://github.com/MaibornWolff/SecObserve/tree/feat/vex_poc contains an initial implementation of the backend to generate VEX documents for CSAF and OpenVEX:

* VEX documents can be generated for a product or a list of vulnerability names (CVE's etc.) or a combination of both.
* VEX documents can be updated. If the vulnerabilities haven't changed, no update is done and the service returns a 204 without a document. Otherwise a complete VEX document is generated with the version number +1.
* Both types of VEX documents should be conform with the respective specification. To test that I used:
  * **CSAF:** The csaf_validator_service (https://github.com/secvisogram/csaf-validator-service)
  * **OpenVEX:** A JSON Schema which I submitted as a PR to the OpenVEX specification (https://github.com/openvex/spec/pull/45)
* A `uuid` is generated as part of the document id, to ensure uniqueness. This `uuid` has to be used as the key to update a document.
* Currently only the user who has generated a VEX document is allowed to update it. I plan to implement user groups in the future, which will make this less restrictive.

What is missing:

* The `justification` for not affected vulnerabilities is not yet implemented.
* `purl` and `cpe23` for products are not yet implemented
* A lot of testing

To test it you have to:
* Set the feature flag `FEATURE_VEX` to `True` in the admin interface (*SECOBSERVE_BACKEND_URL/admin*)
* checkout the branch https://github.com/MaibornWolff/SecObserve/tree/feat/vex_poc and have a look at the Swagger UI (*SECOBSERVE_BACKEND_URL/api/oa3/swagger-ui#/*) to see the new API endpoints:
  * `/api/vex/csaf_document/create/`
  * `/api/vex/csaf_document/update/{document_base_id}/`
  * `/api/vex/openvex_document/create/`
  * `/api/vex/openvex_document/update/{document_base_id}/`

Looking forward to your feedback 😃

**Reply by dervoeti** [reply]
Awesome, will have a look at it this week! The work for "justification" is also almost done, PR will be ready very soon.

**Comment by StefanFl** [comment]
The `feat/vex_poc` branch has now a user interface to create and update VEX documents. The documents contain the justification (thank you so much @dervoeti) and purl's and cpe`s (if entered for the product).

It seems to be a meaningful implemention to me, but I am looking forward very much for your feedback.

**Reply by dervoeti** [reply]
Totally awesome  🎉  I will test this later today. We've been very actively developing our vulnerability management process this week and outlined some ideas. We plan to further refine our ideas next week and then we will provide feedback (also happy to contribute with PRs when it makes sense).

**Comment by StefanFl** [comment]
I made 2 major changes:
* In general the UI now uses the term "Branch / Version" instead only "Branch". Having vulnerability checks by version instead of branch makes a lot of sense, in particular for SCA.
* When you generate a VEX document for a product, you can now select which branches/ versions to include. If no branch / version is selected, all of them will be included. (Before this change it was only the default branch / version.

**Comment by StefanFl** [comment]
Another change: I wasn't happy with the usage of uuid to make the filenames unique. After a look how other CSAF providers are doing it, it is now a combination of prefix, year and a counter.

This should be really it from my side.

**Comment by dervoeti** [comment]
That looks pretty good already, thanks!
We're currently thinking about how to solve the following scenarios:

1. A product version is affected by a vulnerability and there's a mitigation or a workaround available for it. This would map to the CSAF remediation categories `mitigation` and `workaround`.
2. A product version is affected and there's a newer version available that fixes the vulnerability. This would map to the CSAF remediation category `vendor_fix`. In theory this could automatically be determined, if a newer version is present in SecObserve that does not have this vulnerability (that requires some form of version comparison though). Or the user would manually specify that version when selecting `vendor_fix`.

Things to consider:
 - The CSAF `remediation` currently depends on the SecObserve `recommendation`. It looks like you can't specify a `recommendation` for imported observations.
 - As far as I see it, CSAF allows to specify multiple remediations for one observation (e.g. `vendor_fix` and `workaround`, so people can choose whether they want to upgrade or implement the suggested workaround).
 - Would the status "Risk accepted" be the appropiate one for both scenarios?

I'm not sure what's the best way to solve this, this is quite complex. Perhaps users could optionally provide a list of remediations when selecting "Risk accepted" in the assessment dialog. Happy to implement this, but I wanted to discuss the best way to do it first.

**Comment by StefanFl** [comment]
I see what you mean, but have to think about, what the best solution is. I don't think it's the status "Risk accepted", that means it is as it is, I cannot do something at the moment and accept the risk that is involved with the vulnerability.

We might need a list of remediations with different categories, that are editable for imported vulnerabilities. They would need to be concatenated for OpenVEX.

But I wouldn't include that in the current branch. I would like to merge the current branch into dev and release it soon. It is marked as experimental and must explicitely be activated. Then we can go from there with enhancements and changes.

**Comment by dervoeti** [comment]
Yep, totally fine, the current state is already quite useful. Just wanted to mention this since we were thinking about what else we would need in our VEX statements. But I also think that it's a good idea to implement this in a separate step, since it might turn out to be a complicated task.

**Comment by StefanFl** [comment]
I just found a merged PR for Trivy, which amends how they deal with VEX documents: https://github.com/aquasecurity/trivy/pull/5923

It might be a good idea, to generate CSAF documents like their test example https://github.com/juan131/trivy/blob/main/pkg/vex/testdata/csaf-not-affected-sub-components.json. 

Currently in our CSAF documents all vulnerabilities are mapped to the SecObserve product. Maybe it is better to map the vulnerability to its component and the components is related to the product. This would be similar to how we produce OpenVEX documents with products and subcomponents.

What do you think @dervoeti  and @lfrancke?

**Comment by dervoeti** [comment]
I think that's a very good idea and it's probably even necessary to implement it like that. A vulnerability might affect multiple components of a product and you might want to issue different statements for each affected component of the product. The vulnerability might be exploitable in component A but not in component B, even though scanners report the vulnerability in both components. In that case you could provide a workaround for component A and set component B to "Not affected". That's already possible in SecObserve, since you'd have two different observations in that case, but this level of detail needs to be reflected in VEX documents as well.

**Reply by StefanFl** [reply]
The components are now part of the CSAF document and the status of the vulnerabilities are now linked to a relationship of component_version to product_version. This make the CSAF documents much more meaningful. Please checkout the dev branch and test it.

**Reply by dervoeti** [reply]
Just checked this, looks very good already, thanks a lot! At our company, we need to look into how the hierarchy of the product tree should look like. We might have something like "Stackable (vendor) -> Stackable Data Platform (product family) -> 23.11 (release version) -> SDP Zookeeper (product) -> 3.8.3 (product version)". And then probably a list or tree of external components that are part of that product version. But I don't think it's the responsibility of SecObserve to provide an interface for this, as it will be hard/impossible to provide a "one fits all" solution. Our approach would be to take the CSAF document from SecObserve as a basis and then modify the product tree in an automated way to fit our needs.

**Comment by dervoeti** [comment]
> Currently in our CSAF documents all vulnerabilities are mapped to the SecObserve product. Maybe it is better to map the vulnerability to its component and the components is related to the product.

I've been giving this some more thought today, it seems to be more complex than I thought (as always), this is the result: https://github.com/MaibornWolff/SecObserve/discussions/1254
But I don't think this is in any way urgent or necessary, the current solution is fine and might even be the most reasonable one. Just wanted to check the tradeoffs.

**Comment by StefanFl** [comment]
With the new release 1.9.1 not only the generation of CSAF with components (see above) is productive, but there is a change in the authorization. VEX documents that have been created for a product now follows the product authorization. Maintainers and owners of a product can generate, update and delete VEX documents.

**Reply by lfrancke** [reply]
This sounds great, thank you!

Speaking of authorization - and I can create a new issue for this: I think we'll need a workflow system in the future. I'm not sure if this should be in SecObserve or via an integration into an external system. I think I'd prefer it in SecObserve.

What I mean is a approval/review system for vulnerability assessments. User A makes an assessment but before it is "public" it needs to be approved/reviewed by someone else. Just like with code reviews.
Wrong assessments can lead to financial loss/liabilities so we'd like to have some protection in place against mistakes.

We tried various commercial tools and didn't find a single one with this feature.

**Reply by StefanFl** [reply]
Good idea, please write an issue for it. I would make it optional per product. So you can decide for a product if assessments get in effect immediately (like now) or need a review.

**Reply by lfrancke** [reply]
Done: https://github.com/MaibornWolff/SecObserve/issues/1276

**Comment by StefanFl** [comment]
Discussion closed as generation of VEX documents is implemented
